File: src/gui_backend/gui_backend/core/payloads.py
"""Wire payloads.

The contract between the backend and the browser. Kept in one module, as plain
functions over plain data, so that the shape of every message is readable in one
place and testable without a socket.

Two conventions hold everywhere:

* **``source_utc_ms`` is when the value was produced, not when it was sent.**
  Every payload carries it. The client renders data age from it, which is the
  only way an operator on an intermittent link can tell a current position from
  a stale one (safety rule 7).
* **Detail levels shrink the payload, they do not lie about it.** A ``minimal``
  vessel payload has fewer fields; the fields it does have mean exactly what
  they mean at ``full``. Nothing is rounded into uselessness and presented as
  precise.
"""

from __future__ import annotations

import math

from asket_common.heading import HeadingEstimate
from asket_common.mode_arbitration import arming_block_reason
from asket_common.survey import SurveyPlan

from .streams import DETAIL_FULL, DETAIL_MINIMAL, DETAIL_REDUCED


def _f(value: float | None, digits: int = 6) -> float | None:
    """Round for the wire, and turn NaN into ``null``.

    JSON has no NaN. Serialising it produces invalid JSON that some parsers
    accept and others reject; ``null`` means "not available", which is what a
    NaN meant anyway.
    """
    if value is None:
        return None
    if isinstance(value, float) and (math.isnan(value) or math.isinf(value)):
        return None
    return round(float(value), digits)


def _opt_bool(value):
    """``None`` stays ``None``; anything else becomes a real bool.

    ``bool(None)`` is ``False``, and ``False`` here means "the Pico told us it
    is not so". Collapsing the two turns a field that was never sent into a
    confident negative — for ``rc_link_ok`` that reads on screen as "RC lost",
    which is the single worst lie this panel could tell.
    """
    return None if value is None else bool(value)


def _opt_int(value):
    """Same, for counts. ``int(None)`` does not fail quietly — it raises, and
    it raised here the moment a real STATE line reached this function."""
    return None if value is None else int(value)


# -- vessel ---------------------------------------------------------------


def vessel_payload(sample, detail: str = DETAIL_FULL) -> dict:
    """Position and motion.

    Position survives into every detail level, including LTE-M. If the link is
    down to a beacon, where the boat is remains the one thing worth a kilobyte.
    """
    out = {
        "lat": _f(sample.lat, 7),
        "lon": _f(sample.lon, 7),
        "heading_deg": _f(sample.heading_deg, 1),
        "heading_valid": bool(sample.heading_valid),
    }
    if detail == DETAIL_MINIMAL:
        return out

    out.update(
        {
            "cog_deg": _f(sample.cog_deg, 1),
            "sog_ms": _f(sample.sog_ms, 2),
            "heading_source": sample.heading_source,
        }
    )
    if detail == DETAIL_REDUCED:
        return out

    out.update(
        {
            "alt_m": _f(sample.alt, 2),
            "roll_deg": _f(sample.roll_deg, 2),
            "pitch_deg": _f(sample.pitch_deg, 2),
            "gnss_fix_type": _opt_int(sample.gnss_fix_type),
            # Absent, not zero, and therefore not int() either.
            #
            # `vessel_from_odometry` sets this to None on purpose: neither
            # NavSatFix nor Odometry carries a satellite count and this stack
            # has no MAVROS GPSRAW, so "0 satellites" would read as a GNSS
            # failure and ground a healthy vessel. int(None) raises, and it
            # raised on the first real Gazebo fix that reached this function —
            # killing the hub's tick loop and with it every stream, at every
            # detail level, for as long as the page stayed open.
            #
            # `_opt_int` and `_opt_bool` were written for exactly this in
            # `pico_payload` and never applied here. SimSource always supplies
            # a real count, so nothing in sim could reach the branch.
            "num_sats": _opt_int(sample.num_sats),
            "hdop": _f(sample.hdop, 2),
            "distance_travelled_m": _f(sample.distance_travelled_m, 1),
            "on_survey": bool(sample.on_survey),
        }
    )
    return out


def heading_payload(estimate: HeadingEstimate, detail: str = DETAIL_FULL) -> dict:
    """Heading with its provenance.

    Heading error translates directly into seabed position error — about 90 cm
    per degree at 50 m — so the number never travels without its source, its
    validity and how far it has drifted from course over ground.
    """
    out = {
        "heading_deg": _f(estimate.heading_deg, 1),
        "source": estimate.source,
        "valid": bool(estimate.valid),
        "divergence_deg": _f(estimate.divergence_deg, 1),
        "divergence_meaningful": bool(estimate.divergence_meaningful),
        "divergence_suspicious": bool(estimate.divergence_suspicious),
    }
    if detail == DETAIL_FULL:
        out["accuracy_deg"] = _f(estimate.accuracy_deg, 2)
        # Whether the device said so or we assumed it for the class of device.
        # Shown identically otherwise, the two would be indistinguishable, and
        # a compass reporting a degraded accuracy is the case worth seeing.
        out["accuracy_reported"] = bool(estimate.accuracy_reported)
        out["cog_deg"] = _f(estimate.cog_deg, 1)
        out["sog_ms"] = _f(estimate.sog_ms, 2)
        out["seabed_error_at_50m_m"] = _f(estimate.position_error_at_m(50.0), 2)
    return out


# -- vessel systems -------------------------------------------------------

MODE_NAMES = {0: "ESTOP", 1: "MANUAL", 2: "AUTONOMOUS"}


def pico_payload(sample, detail: str = DETAIL_FULL) -> dict:
    """Confirmed vessel state.

    Everything here is what the Pico says the vessel *is* doing. Nothing in this
    payload ever reflects a request (safety rule 4) — with one labelled
    exception, ``mode_requested_auto``, which is carried precisely so the two
    can be told apart on screen.
    """
    out = {
        "mode": MODE_NAMES.get(sample.mode, "UNKNOWN"),
        "armed": _opt_bool(sample.armed),
        "estop_latched": _opt_bool(sample.estop_latched),
        # Carried on every profile, beacon included. A GUI parsing a firmware
        # it does not understand must say so at any bandwidth, because every
        # other field in this payload is then suspect.
        "firmware_version": _opt_int(getattr(sample, "firmware_version", None)),
        "version_mismatch": _opt_bool(getattr(sample, "version_mismatch", None)),
    }
    if detail == DETAIL_MINIMAL:
        return out

    out.update(
        {
            "rc_link_ok": _opt_bool(sample.rc_link_ok),
            "rc_channel8_raw": _opt_int(getattr(sample, "rc_channel8_raw", None)),
            # The comparison is made once, against the firmware's own
            # threshold, in pico_state. The GUI renders the verdict rather than
            # re-deriving it from a number whose scale it would have to know.
            "ch8_asserting_estop": _opt_bool(getattr(sample, "ch8_asserting_estop", None)),
        }
    )
    if detail == DETAIL_REDUCED:
        return out

    out.update(
        {
            "relay_states": [bool(r) for r in sample.relay_states],
            "esc_status": [int(e) for e in sample.esc_status],
            # Observable, never commandable. Shown so an operator can confirm
            # the sovereign path is where they think it is.
            "hardware_killswitch_engaged": _opt_bool(
                getattr(sample, "hardware_killswitch_engaged", None)
            ),
            "rc_channel7_raw": _opt_int(getattr(sample, "rc_channel7_raw", None)),
            # What the Jetson asked for, next to what it got. "I asked for AUTO
            # and it is still MANUAL" then reads as a refusal by channel 8
            # rather than as a lost command.
            "mode_requested_auto": _opt_bool(getattr(sample, "mode_requested_auto", None)),
            # The Pico's view of our heartbeat. Zero while pico_bridge is
            # running points at the cable, not the software.
            "link_live": _opt_bool(getattr(sample, "link_live", None)),
            # The only measurement of radio *quality* in the system. A rising
            # bad count warns of range loss before the link actually drops.
            "sbus_frames_ok": _opt_int(getattr(sample, "sbus_frames_ok", None)),
            "sbus_frames_bad": _opt_int(getattr(sample, "sbus_frames_bad", None)),
            "sbus_failsafe": _opt_bool(getattr(sample, "sbus_failsafe", None)),
            "sbus_frame_lost": _opt_bool(getattr(sample, "sbus_frame_lost", None)),
            # Why the propellers cannot turn, in one sentence, or absent when
            # nothing is blocking them.
            #
            # Arming is channel 7 and nothing else — deliberately not
            # commandable from software, because somebody is physically present
            # to launch this boat and flipping that switch is their consent.
            # The cost is a state that reads as a fault: a mission permitted,
            # engaged and confirmed in the STATE line, and the vessel moves
            # nothing. The reason is computed once, in asket_common, so the
            # panel, the chip and the log cannot drift into three different
            # explanations of the same switch.
            "arming_block": arming_block_reason(
                _opt_bool(sample.armed),
                rc_arm_high=_opt_bool(getattr(sample, "ch7_arm_high", None)),
                estop_latched=_opt_bool(sample.estop_latched),
            ),
        }
    )
    return out


def power_payload(sample, detail: str = DETAIL_FULL, survey_remaining_m: float | None = None,
                  speed_ms: float | None = None) -> dict:
    """Battery, and the comparison that actually matters.

    With one sonar unit covering one side, the survey distance roughly doubles.
    So the useful question is not "how much charge is left" but "is there enough
    to finish", and that comparison is computed here rather than left to the
    operator's arithmetic on a beach.
    """
    out = {
        "state_of_charge": _f(sample.state_of_charge, 3),
        "voltage": _f(sample.voltage, 2),
    }
    if detail == DETAIL_MINIMAL:
        return out

    out.update(
        {
            "current": _f(sample.current, 2),
            "endurance_s": _f(sample.endurance_s, 0),
        }
    )
    if detail == DETAIL_REDUCED:
        return out

    out.update(
        {
            "power_w": _f(sample.power_w, 1),
            "remaining_wh": _f(sample.remaining_wh, 1),
            "consumed_wh": _f(sample.consumed_wh, 1),
        }
    )

    if survey_remaining_m is not None and speed_ms and speed_ms > 0.1:
        survey_time_s = survey_remaining_m / speed_ms
        out["survey_remaining_m"] = _f(survey_remaining_m, 0)
        out["survey_remaining_s"] = _f(survey_time_s, 0)
        endurance = sample.endurance_s
        if endurance and not math.isinf(endurance):
            out["endurance_margin_s"] = _f(endurance - survey_time_s, 0)
            out["can_finish_survey"] = bool(endurance > survey_time_s)
    return out


# -- sensors --------------------------------------------------------------


def sonar_payload(health, detail: str = DETAIL_FULL) -> dict:
    out = {
        "connected": bool(health.connected),
        "actual_ping_rate_hz": _f(health.actual_ping_rate_hz, 2),
        "clock_offset_ms": int(health.clock_offset_ms),
        "clock_ok": bool(health.clock_ok),
        "clock_compromised": bool(health.clock_compromised),
    }
    if detail != DETAIL_FULL:
        return out

    out.update(
        {
            "commanded_ping_rate_hz": _f(health.commanded_ping_rate_hz, 2),
            "ping_rate_ok": bool(health.ping_rate_ok),
            "points_per_ping": int(health.points_per_ping),
            "valid_points_per_ping": int(health.valid_points_per_ping),
            "speed_of_sound": _f(health.speed_of_sound, 1),
            "range_setting_m": _f(health.range_setting_m, 1),
            "gain_setting": int(health.gain_setting),
            "packet_loss_ratio": _f(health.packet_loss_ratio, 4),
            "pitch_deg": _f(health.pitch_deg, 2),
            "roll_deg": _f(health.roll_deg, 2),
            "checksum_errors": int(health.checksum_errors),
            "bytes_discarded": int(health.bytes_discarded),
            "seconds_since_data": _f(health.seconds_since_data, 1),
            # Whether the ping rate is the device's own figure or our
            # measurement of arrivals. Ours also measures the network, so the
            # panel says which it is showing.
            "rate_from_device": bool(health.rate_from_device),
        }
    )
    return out


def lidar_payload(scan, detail: str = DETAIL_FULL, decimation: int = 1) -> dict:
    """Obstacle returns, top-down and vessel-centred.

    Returns are sent as ``[bearing_deg, range_m]`` pairs rather than a dense
    array of nulls: on a 720-beam scanner over open water, most beams return
    nothing, and sending 720 nulls to say so is the kind of waste that sinks a
    4G link.

    Raw and filtered are both sent at full detail because the operator must be
    able to tell whether the sensor or the filter is at fault.
    """

    def pairs(ranges):
        out = []
        for i in range(0, len(ranges), max(1, decimation)):
            r = ranges[i]
            if r is not None:
                out.append([round(scan.angle_min_deg + i * scan.angle_increment_deg, 1),
                            round(r, 2)])
        return out

    nearest = scan.nearest(use_filtered=True)
    payload = {
        "rotation_hz": _f(scan.rotation_hz, 2),
        "points_per_revolution": int(scan.points_per_revolution),
        # How many beams the scanner swept, as against how many came back with
        # anything. Without it, open water and a blind sensor both read "0", and
        # those need very different responses.
        "beams_per_revolution": len(scan.ranges_m),
        "nearest_range_m": _f(nearest[0], 2) if nearest else None,
        "nearest_bearing_deg": _f(nearest[1], 1) if nearest else None,
        "filtered": pairs(scan.filtered_m),
    }
    if detail == DETAIL_FULL:
        payload["raw"] = pairs(scan.ranges_m)
    return payload


def link_payload(sample, profile: str, profile_manual: bool, clients: int,
                 rate_bytes_per_s: float, detail: str = DETAIL_FULL) -> dict:
    out = {
        "active_link": sample.active_link,
        "quality": _f(sample.quality, 2),
        "rtt_ms": _f(sample.rtt_ms, 0),
        "profile": profile,
        "profile_manual": bool(profile_manual),
    }
    if detail == DETAIL_FULL:
        out.update(
            {
                "capacity_bytes_per_s": _f(sample.capacity_bytes_per_s, 0),
                "rate_bytes_per_s": _f(rate_bytes_per_s, 0),
                "connected_clients": int(clients),
                "distance_m": _f(sample.distance_m, 0),
            }
        )
    return out


# -- map layers -----------------------------------------------------------


def track_payload(points: list[tuple[float, float]], cursor: int, step: int = 1) -> dict:
    """The vessel track, as an increment.

    ``from`` is where this increment starts; the client appends. Coordinates go
    out as ``[lon, lat]`` pairs rounded to seven places — about a centimetre,
    which is finer than anything we can actually measure.
    """
    new = points[cursor:]
    if step > 1:
        new = new[::step]
    return {
        "from": cursor,
        "total": len(points),
        "points": [[round(lon, 7), round(lat, 7)] for lat, lon in new],
    }


#: One coverage sample on the wire, as a fixed-order array rather than an
#: object: ``[lat, lon, heading_deg, half_width_m, inner_gap_m]``. A gap in the
#: data is ``null``. Objects cost about ninety bytes a sample and arrays about
#: forty, and over a three-hour survey that difference is megabytes.
COVERAGE_FIELDS = ["lat", "lon", "heading_deg", "half_width_m", "inner_gap_m"]


def coverage_payload(segments: list[dict], cursor: int, side: str, step: int = 1) -> dict:
    """The swath ribbon, as an increment.

    A ``null`` entry is a gap — a stretch where the sonar was not ensonifying,
    because the vessel was turning, the heading was invalid, or the sonar had
    dropped out. The client renders gaps as gaps. With one sonar unit covering
    one side only, a gap that renders as filled is the most expensive way this
    GUI could mislead an operator.
    """
    new = segments[cursor:]
    if step > 1:
        new = new[::step]
    encoded: list[list[float] | None] = []
    for segment in new:
        if segment.get("gap"):
            encoded.append(None)
        else:
            encoded.append([
                round(segment["lat"], 7),
                round(segment["lon"], 7),
                round(segment["heading_deg"], 1),
                round(segment["half_width_m"], 1),
                round(segment["inner_gap_m"], 1),
            ])
    return {"from": cursor, "total": len(segments), "side": side, "segments": encoded}


def plan_payload(plan: SurveyPlan, geofence: list[tuple[float, float]] | None = None) -> dict:
    """Survey lines and the geofence. Sent on subscribe, then on change."""
    return {
        "lines": [
            {"index": i, "coords": [[lon, lat] for lat, lon in seg]}
            for i, seg in enumerate(plan.line_segments_geo())
        ],
        "sonar_side": plan.sonar_side,
        "line_spacing_m": _f(plan.line_spacing_m, 1),
        "total_survey_distance_m": _f(plan.total_survey_distance_m(), 0),
        "geofence": [[lon, lat] for lat, lon in (geofence or [])],
    }

File: src/asket_gui/src/lib/mock/payloads.js
// Wire payloads, mirroring `gui_backend/core/payloads.py`.
//
// Same rules as the Python, for the same reasons:
//
//  * a value that is not available is `null`, never zero — on a boat, a zero is
//    a measurement;
//  * detail levels shrink the payload, they do not lie about it. A `minimal`
//    vessel frame has fewer fields; the fields it has mean exactly what they
//    mean at `full`.
//
// `test_mock_payload_shapes.py` compares the key sets produced here against the
// Python, per stream and per detail level, so a panel cannot come to depend on
// a field the real backend never sends.

const round = (value, digits) => {
  if (value === null || value === undefined || Number.isNaN(value)) return null;
  if (!Number.isFinite(value)) return null;
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
};

export function vesselPayload(world, detail) {
  const { lat, lon } = world.position();
  const out = {
    lat: round(lat, 7),
    lon: round(lon, 7),
    heading_deg: round(world.headingDeg(), 1),
    heading_valid: world.headingValid(),
  };
  if (detail === 'minimal') return out;

  Object.assign(out, {
    cog_deg: round(world.cogDeg(), 1),
    sog_ms: round(world.sogMs(), 2),
    heading_source: world.headingSource(),
  });
  if (detail === 'reduced') return out;

  const { roll, pitch } = world.attitudeDeg();
  const gnss = world.gnss();
  Object.assign(out, {
    alt_m: 0,
    roll_deg: round(roll, 2),
    pitch_deg: round(pitch, 2),
    gnss_fix_type: gnss.fixType,
    num_sats: gnss.numSats,
    hdop: round(gnss.hdop, 2),
    distance_travelled_m: round(world.distanceTravelled, 1),
    on_survey: world.onSurvey(),
  });
  return out;
}

export function headingPayload(world, detail) {
  const heading = world.headingDeg();
  const cog = world.cogDeg();
  const sog = world.sogMs();
  const valid = world.headingValid();
  const meaningful = valid && sog >= 0.6;
  const divergence = valid ? ((heading - cog + 540) % 360) - 180 : null;
  const accuracy = world.headingAccuracyDeg();

  const out = {
    heading_deg: valid ? round(heading, 1) : null,
    source: world.headingSource(),
    valid,
    divergence_deg: round(divergence, 1),
    divergence_meaningful: meaningful,
    divergence_suspicious: Boolean(meaningful && Math.abs(divergence) > 15),
  };
  if (detail === 'full') {
    out.accuracy_deg = round(accuracy, 2);
    // Both simulated sources compute their own figure, so it is a reported one.
    // The flag exists for the real system: when MAVROS reports no accuracy the
    // backend substitutes a nominal one for the class of hardware, and the
    // panel labels that "assumed" rather than showing the two identically.
    out.accuracy_reported = valid && accuracy !== null;
    out.cog_deg = round(cog, 1);
    out.sog_ms = round(sog, 2);
    // The number that makes heading matter: about 90 cm per degree at 50 m.
    out.seabed_error_at_50m_m =
      accuracy === null ? null : round(50 * Math.tan((accuracy * Math.PI) / 180), 2);
  }
  return out;
}

const MODE_NAMES = { 0: 'ESTOP', 1: 'MANUAL', 2: 'AUTONOMOUS' };
const MODE_AUTONOMOUS = 2;

// Ch8 below this is MODE_ESTOP in the firmware. A raw SBUS count, not a
// percentage — mirrors CH8_ESTOP_MAX in world.js, pico.py and pico_state.py.
const CH8_ESTOP_MAX = 700;

// Ch7 above this is "armed" in the firmware. Mirrors ARM_THRESHOLD in the
// sketch and CH7_ARM_MIN in pico_state.py.
const CH7_ARM_MIN = 1000;

// Why the propellers cannot turn. The real backend computes this in
// `asket_common.mode_arbitration.arming_block_reason()` and ships the sentence;
// the mock has to produce the same sentences or a reviewer reads wording in
// mock mode that the boat will never send.
//
// `test_mock_payload_shapes.py` compares these four strings against the Python
// constants, so this is a mirror that cannot drift quietly.
const ARM_BLOCKED_RC_LOW = 'RC channel 7 disarmed, propulsion cannot start';
const ARM_BLOCKED_LATCHED =
  'e-stop latched, propulsion cannot start until the arm switch is cycled';
const ARM_BLOCKED_UNKNOWN = 'the vessel reports disarmed, so propulsion cannot start';
const ARM_BLOCKED_CONTRADICTORY =
  'the vessel reports disarmed although channel 7 is up — propulsion cannot '
  + 'start, and the two disagree';

export function armingBlockReason(armed, rcArmHigh, estopLatched) {
  if (armed === true) return null;
  // Not knowing is not evidence of a block: an unarmed-because-unheard-from
  // vessel gets no red line. This is the state every panel starts in.
  if (armed === null || armed === undefined) return null;
  if (rcArmHigh === false) return ARM_BLOCKED_RC_LOW;
  if (estopLatched === true) return ARM_BLOCKED_LATCHED;
  if (rcArmHigh === true) return ARM_BLOCKED_CONTRADICTORY;
  return ARM_BLOCKED_UNKNOWN;
}

export const ARM_BLOCK_STRINGS = {
  rc_low: ARM_BLOCKED_RC_LOW,
  latched: ARM_BLOCKED_LATCHED,
  unknown: ARM_BLOCKED_UNKNOWN,
  contradictory: ARM_BLOCKED_CONTRADICTORY,
};

export function picoPayload(world, detail) {
  const out = {
    mode: MODE_NAMES[world.mode] || 'UNKNOWN',
    armed: world.armed,
    estop_latched: world.estopLatched,
    // On every profile, beacon included: a GUI reading a firmware it does not
    // understand must be able to say so at any bandwidth, because every other
    // field here is then suspect.
    firmware_version: world.firmwareVersion,
    version_mismatch: false,
  };
  if (detail === 'minimal') return out;

  Object.assign(out, {
    rc_link_ok: world.rcLinkOk,
    rc_channel8_raw: world.rcChannel8,
    ch8_asserting_estop: world.rcChannel8 < CH8_ESTOP_MAX,
  });
  if (detail === 'reduced') return out;

  Object.assign(out, {
    // One relay, matching the firmware and asket_sim. This used to be four,
    // which rendered as `0/4 relays closed` — a false statement about the
    // e-stop path, in the mode the GUI gets reviewed in.
    relay_states: [world.armed],
    esc_status: [world.armed ? 0 : 1, world.armed ? 0 : 1],
    hardware_killswitch_engaged: false,
    rc_channel7_raw: world.rcChannel7,
    mode_requested_auto: world.mode === MODE_AUTONOMOUS,
    link_live: true,
    sbus_frames_ok: world.sbusFramesOk,
    sbus_frames_bad: 0,
    sbus_failsafe: !world.rcLinkOk,
    sbus_frame_lost: false,
    arming_block: armingBlockReason(
      world.armed,
      world.rcChannel7 > CH7_ARM_MIN,
      world.estopLatched,
    ),
  });
  return out;
}

export function powerPayload(world, detail) {
  const capacity = world.cfg.batteryCapacityWh;
  const soc = world.remainingWh / capacity;
  const ocv = 21 + (29.4 - 21) * (0.15 + 0.7 * soc + 0.15 * Math.tanh((soc - 0.5) * 6) / Math.tanh(3));
  const voltage = Math.max(21, Math.min(29.4, ocv));
  const current = world.avgPowerW / Math.max(1, voltage);

  const out = { state_of_charge: round(soc, 3), voltage: round(voltage, 2) };
  if (detail === 'minimal') return out;

  const enduranceS = world.avgPowerW > 1 ? (world.remainingWh / world.avgPowerW) * 3600 : null;
  Object.assign(out, { current: round(current, 2), endurance_s: round(enduranceS, 0) });
  if (detail === 'reduced') return out;

  Object.assign(out, {
    power_w: round(world.avgPowerW, 1),
    remaining_wh: round(world.remainingWh, 1),
    consumed_wh: round(world.consumedWh, 1),
  });

  // The comparison that actually matters: with one sonar covering one side the
  // survey distance roughly doubles, so "is there enough to finish" beats
  // "how much is left".
  const sog = world.sogMs();
  const remainingM = world.surveyRemainingM();
  if (sog > 0.1) {
    const surveyTimeS = remainingM / sog;
    out.survey_remaining_m = round(remainingM, 0);
    out.survey_remaining_s = round(surveyTimeS, 0);
    if (enduranceS !== null) {
      out.endurance_margin_s = round(enduranceS - surveyTimeS, 0);
      out.can_finish_survey = enduranceS > surveyTimeS;
    }
  }
  return out;
}

export function sonarPayload(world, detail) {
  const connected = !world.hasFault('sonar_dropout');
  const offset = Math.round(world.clockOffsetMs);
  const out = {
    connected,
    actual_ping_rate_hz: connected ? round(world.cfg.sonarPingRateHz, 2) : 0,
    clock_offset_ms: offset,
    clock_ok: Math.abs(offset) < 250,
    clock_compromised: Math.abs(offset) >= 1000,
  };
  if (detail !== 'full') return out;

  const { roll, pitch } = world.attitudeDeg();
  Object.assign(out, {
    commanded_ping_rate_hz: round(world.cfg.sonarPingRateHz, 2),
    ping_rate_ok: connected,
    points_per_ping: world.cfg.sonarPointsPerPing,
    valid_points_per_ping: connected ? Math.round(world.cfg.sonarPointsPerPing * 0.8) : 0,
    speed_of_sound: 1500,
    range_setting_m: round(world.cfg.sonarRangeM, 1),
    gain_setting: world.cfg.sonarGain,
    packet_loss_ratio: world.hasFault('sonar_packet_loss') ? 0.12 : 0,
    pitch_deg: round(pitch, 2),
    roll_deg: round(roll, 2),
    checksum_errors: 0,
    bytes_discarded: 0,
    seconds_since_data: connected ? 0 : 30,
    rate_from_device: connected,
  });
  return out;
}

export function lidarPayload(world, detail) {
  const scan = world.lidarScan();
  const step = { full: 1, reduced: 4, minimal: 12 }[detail];
  const decimate = (pairs) => pairs.filter((_, i) => i % step === 0);

  const out = {
    rotation_hz: round(scan.rotationHz, 2),
    points_per_revolution: scan.pointsPerRevolution ?? 0,
    // Beams swept, as against beams that returned something. Open water and a
    // blind sensor both report zero returns; only this tells them apart.
    beams_per_revolution: scan.beamsPerRevolution ?? 0,
    nearest_range_m: round(scan.nearestRangeM, 2),
    nearest_bearing_deg: round(scan.nearestBearingDeg, 1),
    filtered: decimate(scan.filtered),
  };
  // Raw only at full detail: the operator needs both to tell whether the sensor
  // or the filter is at fault, and it doubles the payload.
  if (detail === 'full') out.raw = decimate(scan.raw);
  return out;
}

export function linkPayload(world, detail, context) {
  const link = world.link();
  const out = {
    active_link: link.activeLink,
    quality: round(link.quality, 2),
    rtt_ms: round(link.rttMs, 0),
    profile: context.profile,
    profile_manual: context.manual,
  };
  if (detail === 'full') {
    Object.assign(out, {
      capacity_bytes_per_s: round(link.capacityBytesPerS, 0),
      rate_bytes_per_s: round(context.rateBytesPerS, 0),
      connected_clients: 1,
      distance_m: round(link.distanceM, 0),
      // What the radio sees. Null on a bearer the budget does not model, and
      // null is the honest answer there — a zero would render as "no signal".
      rssi_dbm: round(link.rssiDbm, 1),
      rssi_source: link.rssiSource ?? null,
      expected_rssi_dbm: round(link.expectedRssiDbm, 1),
      headroom_db: round(link.headroomDb, 1),
      mcs_index: link.mcsIndex,
      phy_mbps: round(link.phyMbps, 1),
      // Geometry, which needs no radio telemetry at all, and which is reported
      // even when the link is gone — that is when it is worth having.
      off_boresight_deg: round(link.offBoresightDeg, 1),
      sector_beamwidth_deg: round(link.sectorBeamwidthDeg, 0),
      vessel_off_boresight_deg: round(link.vesselOffBoresightDeg, 1),
      multipath_db: round(link.multipathDb, 1),
      sea_state_m: round(link.seaStateM, 2),
    });
  }
  return out;
}

export function missionPayload(world) {
  const elapsed = world.missionState === 'RECORDING'
    ? (world.utcMs - world.missionStartUtc) / 1000
    : 0;
  return {
    state: world.missionState,
    name: world.missionName,
    mission_dir: world.missionName ? `/data/missions/${world.missionName}` : '',
    elapsed_s: round(elapsed, 1),
    bytes_written: Math.round(world.missionBytes),
    disk_free_bytes: world.diskFreeBytes(),
    disk_total_bytes: world.cfg.diskTotalBytes,
    estimated_remaining_s:
      world.missionState === 'RECORDING'
        ? Math.round(Math.max(0, world.diskFreeBytes() - 1024 ** 3) / (3.2 * 1024 ** 2))
        : null,
    trajectory_records: Math.round(elapsed * 10),
    error_message: world.missionError || '',
    missions: world.missions,
    destinations: world.mockDestinations || [],
    no_fast_path_message:
      'Transfer unavailable — connect a USB drive or an Ethernet cable. Mission data is ' +
      'several gigabytes per hour and will not go over the wireless link.',
  };
}

export function trackPayload(world, cursor, step) {
  const slice = world.track.slice(cursor);
  return {
    from: cursor,
    total: world.track.length,
    points: step > 1 ? slice.filter((_, i) => i % step === 0) : slice,
  };
}

export function coveragePayload(world, cursor, step) {
  const slice = world.coverage.slice(cursor);
  return {
    from: cursor,
    total: world.coverage.length,
    side: world.cfg.sonarSide,
    segments: step > 1 ? slice.filter((_, i) => i % step === 0) : slice,
  };
}
